File: java-gdb-activities-20/activity16-solution/src/com/gdb/tests/TestTransactionModel.java
package com.gdb.tests;

import com.gdb.domain.*;
import com.gdb.service.TransferService;
import com.gdb.exceptions.*;

public class TestTransactionModel {
    public static void main(String[] args) throws Exception {
        System.out.println("=".repeat(60));
        System.out.println("  ACTIVITY 16 — TRANSACTION MODEL TEST");
        System.out.println("=".repeat(60));

        TransferService svc = new TransferService();

        // 📝 STEP 10: Deposit with Transaction
        Account acc1 = (Account) AccountFactory.createAccount("SAVINGS", 1001, "Rajesh Sharma", 30, 50000);
        acc1.setPin(1234);
        Transaction t1 = acc1.depositWithTransaction(5000);
        System.out.println("[STEP 10] Deposit Transaction: " + t1);

        // 📝 STEP 11: Withdraw with Transaction
        Transaction t2 = acc1.withdrawWithTransaction(2000, 1234);
        System.out.println("[STEP 11] Withdrawal Transaction: " + t2);

        // 📝 STEP 12: Transfer with Transaction
        Account acc2 = (Account) AccountFactory.createAccount("SAVINGS", 1002, "Priya Patel", 28, 20000);
        Transaction t3 = svc.transferWithTransaction(acc1, acc2, 1000, 1234);
        System.out.println("[STEP 12] Transfer Transaction: " + t3);

        // 📝 STEP 13: Backward Compatibility Check
        acc1.deposit(1000);
        System.out.println("[STEP 13] Legacy Deposit +1000: " + acc1.getAccountInfo());
    }
}

File: python-gdb-activities-20/activity18/gdb/tests/test_transaction_model.py
# gdb/tests/test_transaction_model.py
import unittest
from datetime import datetime
from gdb.domain.account_factory import AccountFactory
from gdb.domain.transaction import Transaction
from gdb.domain.transaction_type import TransactionType

class TestTransactionModel(unittest.TestCase):
    def setUp(self):
        self.acc = AccountFactory.create_account("SAVINGS", 1001, "Rajesh Sharma", 30, 50000.0, 3)
        self.acc.set_pin(1234)

    def test_deposit_with_transaction(self):
        txn = self.acc.deposit_with_transaction(5000.0)
        self.assertIsNotNone(txn)
        self.assertEqual(txn.get_type(), TransactionType.DEPOSIT)
        self.assertEqual(txn.get_amount(), 5000.0)
        self.assertEqual(txn.get_balance_after(), 55000.0)
        self.assertEqual(txn.get_status(), "SUCCESS")

    def test_withdraw_with_transaction(self):
        txn = self.acc.withdraw_with_transaction(10000.0, 1234)
        self.assertIsNotNone(txn)
        self.assertEqual(txn.get_type(), TransactionType.WITHDRAW)
        self.assertEqual(txn.get_amount(), 10000.0)
        self.assertEqual(txn.get_balance_after(), 40000.0)

    def test_receipt_formatting(self):
        txn = self.acc.deposit_with_transaction(1000.0)
        receipt = txn.get_receipt()
        self.assertIn("TXN ID:", receipt)
        self.assertIn("DEPOSIT", receipt)
        self.assertIn("Rs. 1,000.00", receipt)

if __name__ == "__main__":
    unittest.main()
